Services/JWSProvider.php

```diff
@@ -0,0 +1,192 @@
+<?php
+
+namespace Lexik\Bundle\JWTAuthenticationBundle\Services;
+
+use Lexik\Bundle\JWTAuthenticationBundle\Services\KeyLoader\KeyLoaderInterface;
+use Namshi\JOSE\JWS;
+
+/**
+ * JWS Provider, Namshi\JOSE library integration.
+ * Supports OpenSSL and phpseclib encryption engines.
+ *
+ * @internal
+ *
+ * @since 2.0
+ *
+ * @author Robin Chalas <robin.chalas@gmail.com>
+ */
+class JWSProvider implements JWSProviderInterface
+{
+    /**
+     * @var KeyLoaderInterface
+     */
+    protected $keyLoader;
+
+    /**
+     * @var string
+     */
+    protected $encryptionEngine;
+
+    /**
+     * @var string
+     */
+    protected $encryptionAlgorithm;
+
+    /**
+     * @var JWS
+     */
+    protected $jws;
+
+    /**
+     * @var string
+     */
+    protected $status;
+
+    /**
+     * @param KeyLoaderInterface $keyLoader
+     * @param string             $encryptionEngine
+     * @param string             $encryptionAlgorithm
+     *
+     * @throws \InvalidArgumentException If the given algorithm is not supported.
+     */
+    public function __construct(KeyLoaderInterface $keyLoader, $encryptionEngine, $encryptionAlgorithm)
+    {
+        $encryptionEngine = $encryptionEngine == 'openssl' ? 'OpenSSL' : 'SecLib';
+
+        if (!$this->isAlgorithmSupportedForEngine($encryptionEngine, $encryptionAlgorithm)) {
+            throw new \InvalidArgumentException(
+                sprintf('The algorithm "%s" is not supported for %s', $encryptionAlgorithm, $encryptionEngine)
+            );
+        }
+
+        $this->keyLoader           = $keyLoader;
+        $this->encryptionEngine    = $encryptionEngine;
+        $this->encryptionAlgorithm = $encryptionAlgorithm;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function createSignedToken(array $payload)
+    {
+        $this->jws = new JWS(['alg' => $this->encryptionAlgorithm], $this->encryptionEngine);
+        $this->jws->setPayload($payload);
+        $this->sign();
+
+        return $this;
```

Comment:
I already commented about "state" in this class. So, we need to find a different way to handle this stuff. It looks like you're basing the behavior on the `JWS` class itself, which makes sense - but it's wrong in the same way - it's super-stateful. Ideally, we can "correct" that in our class at least :).

One option is to return a new value object from `createSignedToken` - something with a `getToken()` method and a `getState()` method (or perhaps better, an `isSigned()` method). For `loadSignature()`, you would return another value object with `getStatus()` (or better ,`isVerified()`), `isExpired()` and `getPayload()`. The `JWSProviderInterface` would then only have `createSignedToken()` and `loadSignature()` (but you could re-name these after the update, if there is a better name then).